### PR TITLE
Web editor for repository files (Create new, edit, upload, rename, delete)

### LINKED ISSUE #1071: Web editor for repository files
so that folks can easily make edits to files in a git repository from their browser without having to clone a whole repo

### PR DESCRIPTION
**Pull request by richmahn**
This is a new PR for Issue #1071 which doesn't include all the SimpleMD/CodeMirror files for ease of code review. This PR is dependent upon PR #3286 - "Added coding style modes to SimpleMDE"

This Feature adds the following functionality and improvements:

  * Can edit text files of a repo you have write access to
  * Can create new text files in a repo you have write access to
  * Can upload files to a repo you have write access to (uploading feature has been updated for issues as well)
  * Can delete files of a repo you have write access to
  * Can create a new branch of a repo you have write access to 
  * Uses SimpleMDE/CodeMirror for editing both Markdown files and other coding styles (requires PR #3286)
  * Allows creating sub-directories and renaming files when editing or uploading files
  * Adds the following settings to the app.ini file:
    - Under [upload], can set the number of files that can be uploaded, types of files that can be uploaded, and the max size
    - Under [editor] can set the file extensions that will be line wrapped (including files with no extension), and the types of files that will have a preview tab (such as markdown, yaml, etc)
    - Under [markdown] can set the file extensions of what files will be rendered as markdown, where even no extension is possible

### CODE REVIEW

File: cmd/web.go [hunk 1]
```diff
@@ -489,6 +489,20 @@ func runWeb(ctx *cli.Context) error {
 
 		m.Combo("/compare/*", repo.MustAllowPulls).Get(repo.CompareAndPullRequest).
 			Post(bindIgnErr(auth.CreateIssueForm{}), repo.CompareAndPullRequestPost)
+
```
Comment (richmahn) [inline cmd/web.go]: All the routes for creating/editing/uploading files to a repo and to make new branches

File: conf/app.ini [hunk 2]
```diff
@@ -20,6 +20,26 @@ MAX_CREATION_LIMIT = -1
 ; Patch test queue length, make it as large as possible
 PULL_REQUEST_QUEUE_LENGTH = 10000
 
+[editor]
+; List of file extensions that should have line wraps in the CodeMirror editor
+; Separate extensions with a comma. To line wrap files w/o extension, just put a comma
+LINE_WRAP_EXTENSIONS = .txt,.md,.markdown,.mdown,.mkd,,
+; Valid file modes that have a preview API associated with them, such as api/v1/markdown
+; Separate values by commas. Preview tab in edit mode won't show if the file extension doesn't match
+PREVIEW_TAB_APIS = markdown
```
Comment (richmahn) [inline conf/app.ini]: This is so we can add other previewers, which my company wants, such as converting a .yaml table to an HTML table, thus the API can be added on if one wants. This isn't by file extension, but the name of the code language as determined by CodeMirror.

File: conf/locale/locale_en-US.ini [hunk 3]
```diff
@@ -186,6 +186,13 @@ TeamName = Team name
 AuthName = Authorization name
 AdminEmail = Admin email
 
+NewBranchName = New branch name
```
Comment (richmahn) [inline conf/locale/locale_en-US.ini]: I have no idea why these are camel case and not lower case with underscores....I usually always do the latter.

File: models/repo.go [hunk 4]
```diff
@@ -406,16 +409,25 @@ func (repo *Repository) LocalCopyPath() string {
 	return path.Join(setting.AppDataPath, "tmp/local", com.ToStr(repo.ID))
 }
 
-func updateLocalCopy(repoPath, localPath string) error {
```
Comment (richmahn) [inline models/repo.go]: Needed to add branch name so we could update/create any branch.

File: models/wiki.go [hunk 5]
```diff
@@ -21,44 +21,6 @@ import (
 	"github.com/gogits/gogs/modules/setting"
 )
 
-// workingPool represents a pool of working status which makes sure
```
Comment (richmahn) [inline models/wiki.go]: This was moved out of wiki so that any repo, not just wiki repos, could be have a workingPool. It was moved to working_pool.go for more global use.

File: models/wiki.go [hunk 6]
```diff
@@ -115,7 +77,7 @@ func (repo *Repository) LocalWikiPath() string {
 
 // UpdateLocalWiki makes sure the local copy of repository wiki is up-to-date.
 func (repo *Repository) UpdateLocalWiki() error {
-	return updateLocalCopy(repo.WikiPath(), repo.LocalWikiPath())
+	return updateLocalCopy(repo.WikiPath(), repo.LocalWikiPath(), "")
```
Comment (richmahn) [inline models/wiki.go]: No branch given when creating a wiki repo.

File: modules/markdown/markdown.go [hunk 7]
```diff
@@ -53,10 +53,11 @@ func isLink(link []byte) bool {
 // IsMarkdownFile reports whether name looks like a Markdown file
 // based on its extension.
 func IsMarkdownFile(name string) bool {
-	name = strings.ToLower(name)
```
Comment (richmahn) [inline modules/markdown/markdown.go]: This now uses the markdown file extensions set in app.ini. This is so you can assume .txt or files with no extensions, for example, can be rendered as Markdown on a personal basis.

File: modules/setting/setting.go [hunk 8]
```diff
@@ -437,6 +451,16 @@ func NewContext() {
 		log.Fatal(4, "Fail to map Repository settings: %v", err)
 	}
 
+	sec = Cfg.Section("upload")
+	UploadTempPath = sec.Key("UPLOAD_TEMP_PATH").MustString(path.Join(AppDataPath, "tmp/uploads"))
+	if !filepath.IsAbs(UploadTempPath) {
+		UploadTempPath = path.Join(workDir, UploadTempPath)
+	}
+	UploadAllowedTypes = strings.Replace(sec.Key("UPLOAD_ALLOWED_TYPES").MustString(""), "|", ",", -1)
```
Comment (richmahn) [inline modules/setting/setting.go]: This maybe should be put into an "Upload" struct.

File: public/css/gogs.css [hunk 9]
```diff
@@ -1269,55 +1272,55 @@ footer .ui.language .menu {
 .repository.file.list #file-content .view-raw img {
   padding: 5px 5px 0 5px;
 }
-.repository.file.list #file-content .code-view * {
```
Comment (richmahn) [inline public/css/gogs.css]: Had to make this less restricted so could be used in Wikis and Issue editors.

File: models/issue.go [hunk 10]
```diff
@@ -383,23 +383,18 @@ func newIssue(e *xorm.Session, repo *Repository, issue *Issue, labelIDs []int64,
 	}
 
 	// Check attachments.
-	attachments := make([]*Attachment, 0, len(uuids))
```
Comment (richmahn) [inline models/issue.go]: I changed issues to match what I did for uploading files for repos, so that they had the same variables and settings as repos did.

**Comment by richmahn** [comment]
@Unknwon Ok, have updated this to use the changes in PR #3286 for the CodeMirror modes addon, as well as a fix for deleting files.

File: conf/locale/locale_en-US.ini [hunk 11]
```diff
@@ -801,7 +854,7 @@ teams.read_permission_desc = This team grants <strong>Read</strong> access: memb
 teams.write_permission_desc = This team grants <strong>Write</strong> access: members can read from and push to the team's repositories.
 teams.admin_permission_desc = This team grants <strong>Admin</strong> access: members can read from, push to, and add collaborators to the team's repositories.
 teams.repositories = Team Repositories
-teams.search_repo_placeholder = Search repository...
```
Comment (Unknwon) [inline conf/locale/locale_en-US.ini]: Why change this dots?

File: cmd/web.go [hunk 12]
```diff
@@ -489,6 +489,20 @@ func runWeb(ctx *cli.Context) error {
...
+		m.Group("", func() {
+			m.Combo("/edit/*").Get(repo.EditFile).
+				Post(bindIgnErr(auth.EditRepoFileForm{}), repo.EditFilePost)
+			m.Combo("/new/*").Get(repo.EditNewFile).
```
Comment (Unknwon) [inline cmd/web.go]: Maybe we should use `_edit` and `_new` which same as wiki.

File: cmd/web.go [hunk 13]
```diff
@@ -489,6 +489,20 @@ func runWeb(ctx *cli.Context) error {
...
+				Post(bindIgnErr(auth.EditRepoFileForm{}), repo.EditNewFilePost)
+			m.Post("/preview/*", bindIgnErr(auth.EditPreviewDiffForm{}), repo.DiffPreviewPost)
+			m.Combo("/upload/*").Get(repo.UploadFile).
+				Post(bindIgnErr(auth.UploadRepoFileForm{}), repo.UploadFilePost)
+			m.Post("/delete/*", bindIgnErr(auth.DeleteRepoFileForm{}), repo.DeleteFilePost)
+			m.Post("/branches", bindIgnErr(auth.NewBranchForm{}), repo.NewBranchPost)
+			m.Post("/upload-file", repo.UploadFileToServer)
+			m.Post("/upload-remove", bindIgnErr(auth.UploadRemoveFileForm{}), repo.UploadRemoveFileFromServer)
```
Comment (Unknwon) [inline cmd/web.go]: `UploadRemoveFile*` -> `RemoveUploadedFile*`

File: conf/app.ini [hunk 14]
```diff
@@ -20,6 +20,26 @@ MAX_CREATION_LIMIT = -1
 ; Patch test queue length, make it as large as possible
 PULL_REQUEST_QUEUE_LENGTH = 10000
 
+[editor]
+; List of file extensions that should have line wraps in the CodeMirror editor
+; Separate extensions with a comma. To line wrap files w/o extension, just put a comma
+LINE_WRAP_EXTENSIONS = .txt,.md,.markdown,.mdown,.mkd,,
```
Comment (Unknwon) [inline conf/app.ini]: The last `,` is redundant?

File: conf/app.ini [hunk 15]
```diff
@@ -20,6 +20,26 @@ MAX_CREATION_LIMIT = -1
...
+; Valid file modes that have a preview API associated with them, such as api/v1/markdown
+; Separate values by commas. Preview tab in edit mode won't show if the file extension doesn't match
+PREVIEW_TAB_APIS = markdown
+
+[upload]
```
Comment (Unknwon) [inline conf/app.ini]: How do we deal with this section with `[attachment]`? https://github.com/gogits/gogs/blob/master/conf/app.ini#L230

File: models/error.go [hunk 16]
```diff
@@ -404,6 +404,19 @@ func (err ErrReleaseNotExist) Error() string {
...
+	FileName string
+}
+
+func IsErrRepoFileAlreadyExist(err error) bool {
+	_, ok := err.(ErrRepoFileAlreadyExist)
+	return ok
+}
+
+func (err ErrRepoFileAlreadyExist) Error() string {
+	return fmt.Sprintf("Repo file already exists [file name: %s]", err.FileName)
```
Comment (Unknwon) [inline models/error.go]: `file name` -> `file_name`

File: models/issue.go [hunk 17]
```diff
@@ -430,6 +425,7 @@ func NewIssue(repo *Repository, issue *Issue, labelIDs []int64, uuids []string)
 	act := &Action{
 		ActUserID:    issue.Poster.Id,
 		ActUserName:  issue.Poster.Name,
+
```
Comment (Unknwon) [inline models/issue.go]: What happened to this empty line?

File: models/repo.go [hunk 18]
```diff
@@ -29,12 +29,15 @@ import (
...
 	api "github.com/gogits/go-gogs-client"
+	gouuid "github.com/satori/go.uuid"
 
 	"github.com/gogits/gogs/modules/bindata"
 	"github.com/gogits/gogs/modules/log"
 	"github.com/gogits/gogs/modules/markdown"
 	"github.com/gogits/gogs/modules/process"
 	"github.com/gogits/gogs/modules/setting"
+	"mime/multipart"
+	"io"
```
Comment (Unknwon) [inline models/repo.go]: Please adjust import ordering and grouping reference to https://github.com/Unknwon/go-code-convention/blob/master/en-US/import_packages.md

File: models/repo.go [hunk 19]
```diff
@@ -2099,3 +2111,433 @@ func (repo *Repository) GetForks() ([]*Repository, error) {
...
+}
+
+// CheckoutNewBranch checks out a new branch from the given branch name
+func (repo *Repository) CheckoutNewBranch(oldBranchName, newBranchName string) error {
+	return checkoutNewBranch(repo.RepoPath(), repo.LocalRepoPath(), oldBranchName, newBranchName)
+}
+
+func checkoutNewBranch(repoPath, localPath, oldBranch, newBranch string) error {
+	if !com.IsExist(localPath) {
+		if error := updateLocalCopy(repoPath, localPath, oldBranch); error != nil {
```
Comment (Unknwon) [inline models/repo.go]: `error` -> `err`.

File: models/repo.go [hunk 20]
```diff
@@ -2099,3 +2111,433 @@ func (repo *Repository) GetForks() ([]*Repository, error) {
...
+	repoWorkingPool.CheckIn(com.ToStr(repo.ID))
+	defer repoWorkingPool.CheckOut(com.ToStr(repo.ID))
+
+	if err = repo.DiscardLocalRepoChanges(oldBranchName); err != nil {
+		return fmt.Errorf("discardLocalRepoChanges: %s - %v", oldBranchName, err)
+	} else if err = repo.UpdateLocalRepo(oldBranchName); err != nil {
+		return fmt.Errorf("UpdateLocalRepo: %s - %v", oldBranchName, err)
+	}
+
+	if( oldBranchName != branchName ){
```
Comment (Unknwon) [inline models/repo.go]: Does not seem like you have ran `go fmt`?

File: models/repo.go [hunk 21]
```diff
@@ -2099,3 +2111,433 @@ func (repo *Repository) GetForks() ([]*Repository, error) {
...
+	if len(message) == 0 {
+		if isNewFile {
+			message = "Add '" + treeName + "'"
+		} else {
+			message = "Update '" + treeName + "'"
+		}
+	}
+
+	if ! com.IsExist(filepath.Dir(filePath)) {
+		os.MkdirAll(filepath.Dir(filePath), os.ModePerm)
```
Comment (Unknwon) [inline models/repo.go]: `os.MkdirAll` will do nothing if directory already exists, so we don't need to check here ourselves.

File: models/repo.go [hunk 22]
```diff
@@ -2099,3 +2111,433 @@ func (repo *Repository) GetForks() ([]*Repository, error) {
...
+	}
+
+	if( oldBranchName != branchName ){
+		repo.CheckoutNewBranch(oldBranchName, branchName)
+	}
+
+	dirPath := path.Join(localPath, treeName)
+
+	if ! com.IsExist(dirPath) {
+		os.MkdirAll(dirPath, os.ModePerm)
```
Comment (Unknwon) [inline models/repo.go]: Same as above.

File: models/repo.go [hunk 23]
```diff
@@ -2099,3 +2111,433 @@ func (repo *Repository) GetForks() ([]*Repository, error) {
...
+	} else if err = repo.UpdateLocalRepo(oldBranchName); err != nil {
+		return fmt.Errorf("UpdateLocalRepo: %v", err)
+	}
+
+	if( oldBranchName != branchName ){
+		repo.CheckoutNewBranch(oldBranchName, branchName)
+	}
+
+	dirPath := path.Join(localPath, treeName)
+
```
Comment (Unknwon) [inline models/repo.go]: Remove this line.

File: models/repo.go [hunk 24]
```diff
@@ -2099,3 +2111,433 @@ func (repo *Repository) GetForks() ([]*Repository, error) {
...
+		Name:   name,
+		UID: userId,
+		RepoID: repoId,
+	}
+
+	if err = os.MkdirAll(path.Dir(up.LocalPath()), os.ModePerm); err != nil {
+		return nil, fmt.Errorf("MkdirAll: %v", err)
+	}
+
+	fw, err := os.Create(up.LocalPath())
```
Comment (Unknwon) [inline models/repo.go]: What is the reason not using `io/ioutil.WriteFile`?

File: models/repo.go [hunk 25]
```diff
@@ -2099,3 +2111,433 @@ func (repo *Repository) GetForks() ([]*Repository, error) {
...
+	} else if _, err = io.Copy(fw, file); err != nil {
+		return nil, fmt.Errorf("Copy: %v", err)
+	}
+
+	sess := x.NewSession()
+	defer sessionRelease(sess)
+	if err := sess.Begin(); err != nil {
+		return nil, err
+	}
+	if _, err := sess.Insert(up); err != nil {
```
Comment (Unknwon) [inline models/repo.go]: Only one insert operation does not need session to rollback, just use `x.Insert`

File: models/repo.go [hunk 26]
```diff
@@ -2099,3 +2111,433 @@ func (repo *Repository) GetForks() ([]*Repository, error) {
...
+		return nil, err
+	}
+
+	return up, sess.Commit()
+}
+
+// RemoveUpload removes the file by UUID
+func RemoveUpload(uuid string, userId, repoId int64) (err error) {
+	sess := x.NewSession()
```
Comment (Unknwon) [inline models/repo.go]: Unused session.

File: models/repo.go [hunk 27]
```diff
@@ -2099,3 +2111,433 @@ func (repo *Repository) GetForks() ([]*Repository, error) {
...
+	}
+
+	if err := DeleteUpload(upload, true); err != nil {
+		return fmt.Errorf("DeleteUpload[%s]: %v", uuid, err)
+	}
+
+	return nil
+}
+
+func getUpload(uuid string, userId, repoId int64) (*Upload, error) {
```
Comment (Unknwon) [inline models/repo.go]: All `Id` should be `ID`, reference to https://github.com/Unknwon/go-code-convention/blob/master/en-US/naming_rules.md#variable-naming-convention

File: models/wiki.go (hunk 6 above)
Comment (Unknwon) [inline models/wiki.go]: Should we force to use `master` as some kind of consistency?

**Comment by Unknwon** [comment]
Reviewed few files, please help fix.

Also would you mind give me push access to `richmahn:feature-create-and-edit-repo-file` so I can do some quick update as well.

File: conf/app.ini (hunk 14 above)
Comment (richmahn) [inline conf/app.ini]: Actually it as meaning, as I wrote in the comment before it. I found that usually when a person has a file without a file extension, which a comma that succeeds nothing indicates, it is just a text file, usually a README file or similar. I usually wordwrap those.

Comment (Unknwon) [inline conf/app.ini]: The result of `len(strings.Split(".txt,.md,.markdown,.mdown,.mkd,", ","))` is 6

File: models/wiki.go (hunk 6 above)
Comment (richmahn) [inline models/wiki.go]: I at one time had this as "master" for the wiki, but it wasn't working, found I couldn't specify a branch. Not sure why. If you know how we can force the wiki to have a "master" branch, please let me know.

File: conf/locale/locale_en-US.ini (hunk 11 above)
Comment (richmahn) [inline conf/locale/locale_en-US.ini]: I think my IDE did that. Will change them back.

File: cmd/web.go (hunk 12 above)
Comment (richmahn) [inline cmd/web.go]: Will do.

File: conf/app.ini (hunk 14 above)
Comment (richmahn) [inline conf/app.ini]: OK, good point...I thought the two commas made it more clear, but right, doesn't need two.

File: conf/app.ini (hunk 15 above)
Comment (richmahn) [inline conf/app.ini]: Not sure what you mean here. [attachment] is actually just for issues, and I didn't touch that, so that can be set differently than repository uploads. For example, in a repo, you may want allow any file type to be upload, but for Issues you may just want PDFs and images. If [attachment] should be changed to [issues] that should be a separate pull request.

File: models/repo.go (hunk 23 above)
Comment (richmahn) [inline models/repo.go]: I don't see what line your talking about. 2348 is needed to define dirPath.

File: models/repo.go (hunk 24 above)
Comment (richmahn) [inline models/repo.go]: I did it exactly as models/issue.go does in the NewAttachment() menthod. Was just a copy of that.

File: models/repo.go (hunk 25 above)
Comment (richmahn) [inline models/repo.go]: Ok, but was also done by another developer in models/issue.go at line 1408. Will try 'x'.

File: models/repo.go (hunk 26 above)
Comment (richmahn) [inline models/repo.go]: But this function, taken from issue.go requires a return of what was returned from sess.Commit(). I'm going to leave it in because I had no idea what the developer of issue.go had in mind when returning the session.Commit().

**Comment by richmahn** [comment]
@Unknwon Well, I got to the end of your list making changes, and now I see your whole thing about wanting access and then making your own branch/fork. How should we proceed? I made changes to the fork/branch as noted above.

**Comment by richmahn** [comment]
@Unknwon I don't see a branch related to mine. Where have you gotten with that? I've pushed my changes to this branch. The only thing I didn't change was the Session stuff you commented on, and that wasn't written by me, just copied.

**Comment by richmahn** [comment]
@Unknwon Oh wait, I think I understand now. You were going to push to this branch when you had something. Ok. Use my latest commits if you make changes. Let me know how it goes.

File: models/repo.go (hunk 24 above)
Comment (richmahn) [inline models/repo.go]: https://github.com/gogits/gogs/blob/master/models/issue.go#L1395

File: models/repo.go (hunk 23 above)
Comment (Unknwon) [inline models/repo.go]: 2349.

Comment (richmahn) [inline models/repo.go]: Ok, removed the blank line.

File: conf/app.ini (hunk 15 above)
Comment (Unknwon) [inline conf/app.ini]: OK, maybe it's more clear to rename section to `repository.upload`?

**Comment by rugk** [comment]
It would be nice of you could also support [Editorconfig](http://editorconfig.org/) - just as Github does.
If this is too much for this PR I can also create a new issue for this.

**Comment by andreynering** [comment]
@rugk I did #3241 for Editorconfig support.

I plan to update that when this PR get merged.

Feel free try it and comment, if you want.

**Comment by Unknwon** [comment]
Please sign the CLA.

**Comment by liebo** [comment]
HI, In previous PR https://github.com/gogits/gogs/pull/3151，web editor will change unix newline \n to windows CRLF，resulting shell script not run correctly. any fix of this in new version?

**Comment by Unknwon** [comment]
I have merged commits into branch [feature/web-editor](https://github.com/gogits/gogs/tree/feature/web-editor). 

@richmahn you can continue PR to that branch, and I'll directly work on the branch to test changes and solve conflicts as well.

**Comment by richmahn** [comment]
@Unknwon Ok, got it. Thanks for setting up that branch!

**Comment by richmahn** [comment]
@liebo I do remember this being an issue, but I have no clue to why it is happening. Did you test this branch on a Windows machine by chance?

**Comment by richmahn** [comment]
@Unknwon Do I still need to sign a CLA? How do I do that?

**Comment by Unknwon** [comment]
No you don't, CLA is removed.
